# Backport of nginx commit 4d32a2703c79, Access log: buffer overrun protection.
# Preserve OpenResty log_escape_non_ascii and its lcf argument in escaped length/copy operations.
# The unmodified upstream patch is retained in ../upstream-patches/.
--- a/src/http/modules/ngx_http_log_module.c
+++ b/src/http/modules/ngx_http_log_module.c
@@ -17,7 +17,7 @@
 typedef struct ngx_http_log_op_s  ngx_http_log_op_t;
 
 typedef u_char *(*ngx_http_log_op_run_pt) (ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 
 typedef size_t (*ngx_http_log_op_getlen_pt) (ngx_http_request_t *r,
     uintptr_t data);
@@ -114,40 +114,44 @@
 static void ngx_http_log_flush_handler(ngx_event_t *ev);
 
 static u_char *ngx_http_log_pipe(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_time(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_iso8601(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_msec(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_request_time(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_status(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_bytes_sent(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_body_bytes_sent(ngx_http_request_t *r,
-    u_char *buf, ngx_http_log_op_t *op);
+    u_char *buf, u_char *end, ngx_http_log_op_t *op);
 static u_char *ngx_http_log_request_length(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 
 static ngx_int_t ngx_http_log_variable_compile(ngx_conf_t *cf,
     ngx_http_log_op_t *op, ngx_str_t *value, ngx_uint_t escape);
 static size_t ngx_http_log_variable_getlen(ngx_http_request_t *r,
     uintptr_t data);
 static u_char *ngx_http_log_variable(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static uintptr_t ngx_http_log_escape(ngx_http_log_loc_conf_t *lcf, u_char *dst,
     u_char *src, size_t size);
 static size_t ngx_http_log_json_variable_getlen(ngx_http_request_t *r,
     uintptr_t data);
 static u_char *ngx_http_log_json_variable(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op);
+    u_char *end, ngx_http_log_op_t *op);
 static size_t ngx_http_log_unescaped_variable_getlen(ngx_http_request_t *r,
     uintptr_t data);
 static u_char *ngx_http_log_unescaped_variable(ngx_http_request_t *r,
-    u_char *buf, ngx_http_log_op_t *op);
+    u_char *buf, u_char *end, ngx_http_log_op_t *op);
+
+
+static ngx_int_t ngx_http_log_check_length(ngx_http_request_t *r,
+    u_char *buf, u_char *end, size_t len);
 
 
 static void *ngx_http_log_create_main_conf(ngx_conf_t *cf);
@@ -263,7 +267,7 @@
 static ngx_int_t
 ngx_http_log_handler(ngx_http_request_t *r)
 {
-    u_char                   *line, *p;
+    u_char                   *line, *p, *end;
     size_t                    len, size;
     ssize_t                   n;
     ngx_str_t                 val;
@@ -319,6 +323,8 @@
             }
         }
 
+        len += NGX_LINEFEED_SIZE;
+
         if (log[l].syslog_peer) {
 
             /* length of syslog's PRI and HEADER message parts */
@@ -329,8 +335,6 @@
             goto alloc_line;
         }
 
-        len += NGX_LINEFEED_SIZE;
-
         buffer = log[l].file ? log[l].file->data : NULL;
 
         if (buffer) {
@@ -346,13 +350,18 @@
             if (len <= (size_t) (buffer->last - buffer->pos)) {
 
                 p = buffer->pos;
+                end = p + len - NGX_LINEFEED_SIZE;
 
                 if (buffer->event && p == buffer->start) {
                     ngx_add_timer(buffer->event, buffer->flush);
                 }
 
-                for (i = 0; i < log[l].format->ops->nelts; i++) {
-                    p = op[i].run(r, p, &op[i]);
+                for (i = 0; i < log[l].format->ops->nelts && p; i++) {
+                    p = op[i].run(r, p, end, &op[i]);
+                }
+
+                if (p == NULL) {
+                    return NGX_ERROR;
                 }
 
                 ngx_linefeed(p);
@@ -375,13 +384,18 @@
         }
 
         p = line;
+        end = line + len - NGX_LINEFEED_SIZE;
 
         if (log[l].syslog_peer) {
             p = ngx_syslog_add_header(log[l].syslog_peer, line);
         }
 
-        for (i = 0; i < log[l].format->ops->nelts; i++) {
-            p = op[i].run(r, p, &op[i]);
+        for (i = 0; i < log[l].format->ops->nelts && p; i++) {
+            p = op[i].run(r, p, end, &op[i]);
+        }
+
+        if (p == NULL) {
+            return NGX_ERROR;
         }
 
         if (log[l].syslog_peer) {
@@ -780,7 +794,7 @@
 
 
 static u_char *
-ngx_http_log_copy_short(ngx_http_request_t *r, u_char *buf,
+ngx_http_log_copy_short(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
     size_t     len;
@@ -788,6 +802,10 @@
 
     len = op->len;
     data = op->data;
+
+    if (ngx_http_log_check_length(r, buf, end, len) != NGX_OK) {
+        return NULL;
+    }
 
     while (len--) {
         *buf++ = (u_char) (data & 0xff);
@@ -799,16 +817,25 @@
 
 
 static u_char *
-ngx_http_log_copy_long(ngx_http_request_t *r, u_char *buf,
+ngx_http_log_copy_long(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
+    if (ngx_http_log_check_length(r, buf, end, op->len) != NGX_OK) {
+        return NULL;
+    }
+
     return ngx_cpymem(buf, (u_char *) op->data, op->len);
 }
 
 
 static u_char *
-ngx_http_log_pipe(ngx_http_request_t *r, u_char *buf, ngx_http_log_op_t *op)
-{
+ngx_http_log_pipe(ngx_http_request_t *r, u_char *buf, u_char *end,
+    ngx_http_log_op_t *op)
+{
+    if (ngx_http_log_check_length(r, buf, end, 1) != NGX_OK) {
+        return NULL;
+    }
+
     if (r->pipeline) {
         *buf = 'p';
     } else {
@@ -820,24 +847,43 @@
 
 
 static u_char *
-ngx_http_log_time(ngx_http_request_t *r, u_char *buf, ngx_http_log_op_t *op)
-{
+ngx_http_log_time(ngx_http_request_t *r, u_char *buf, u_char *end,
+    ngx_http_log_op_t *op)
+{
+    if (ngx_http_log_check_length(r, buf, end, ngx_cached_http_log_time.len)
+        != NGX_OK)
+    {
+        return NULL;
+    }
+
     return ngx_cpymem(buf, ngx_cached_http_log_time.data,
                       ngx_cached_http_log_time.len);
 }
 
 static u_char *
-ngx_http_log_iso8601(ngx_http_request_t *r, u_char *buf, ngx_http_log_op_t *op)
-{
+ngx_http_log_iso8601(ngx_http_request_t *r, u_char *buf, u_char *end,
+    ngx_http_log_op_t *op)
+{
+    if (ngx_http_log_check_length(r, buf, end, ngx_cached_http_log_iso8601.len)
+        != NGX_OK)
+    {
+        return NULL;
+    }
+
     return ngx_cpymem(buf, ngx_cached_http_log_iso8601.data,
                       ngx_cached_http_log_iso8601.len);
 }
 
 static u_char *
-ngx_http_log_msec(ngx_http_request_t *r, u_char *buf, ngx_http_log_op_t *op)
+ngx_http_log_msec(ngx_http_request_t *r, u_char *buf, u_char *end,
+    ngx_http_log_op_t *op)
 {
     ngx_time_t  *tp;
 
+    if (ngx_http_log_check_length(r, buf, end, NGX_TIME_T_LEN + 4) != NGX_OK) {
+        return NULL;
+    }
+
     tp = ngx_timeofday();
 
     return ngx_sprintf(buf, "%T.%03M", tp->sec, tp->msec);
@@ -845,11 +891,15 @@
 
 
 static u_char *
-ngx_http_log_request_time(ngx_http_request_t *r, u_char *buf,
+ngx_http_log_request_time(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
     ngx_time_t      *tp;
     ngx_msec_int_t   ms;
+
+    if (ngx_http_log_check_length(r, buf, end, NGX_TIME_T_LEN + 4) != NGX_OK) {
+        return NULL;
+    }
 
     tp = ngx_timeofday();
 
@@ -862,9 +912,14 @@
 
 
 static u_char *
-ngx_http_log_status(ngx_http_request_t *r, u_char *buf, ngx_http_log_op_t *op)
+ngx_http_log_status(ngx_http_request_t *r, u_char *buf, u_char *end,
+    ngx_http_log_op_t *op)
 {
     ngx_uint_t  status;
+
+    if (ngx_http_log_check_length(r, buf, end, NGX_INT_T_LEN) != NGX_OK) {
+        return NULL;
+    }
 
     if (r->err_status) {
         status = r->err_status;
@@ -884,9 +939,13 @@
 
 
 static u_char *
-ngx_http_log_bytes_sent(ngx_http_request_t *r, u_char *buf,
+ngx_http_log_bytes_sent(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
+    if (ngx_http_log_check_length(r, buf, end, NGX_OFF_T_LEN) != NGX_OK) {
+        return NULL;
+    }
+
     return ngx_sprintf(buf, "%O", r->connection->sent);
 }
 
@@ -897,10 +956,14 @@
  */
 
 static u_char *
-ngx_http_log_body_bytes_sent(ngx_http_request_t *r, u_char *buf,
+ngx_http_log_body_bytes_sent(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
     off_t  length;
+
+    if (ngx_http_log_check_length(r, buf, end, NGX_OFF_T_LEN) != NGX_OK) {
+        return NULL;
+    }
 
     length = r->connection->sent - r->header_size;
 
@@ -915,9 +978,13 @@
 
 
 static u_char *
-ngx_http_log_request_length(ngx_http_request_t *r, u_char *buf,
+ngx_http_log_request_length(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
+    if (ngx_http_log_check_length(r, buf, end, NGX_OFF_T_LEN) != NGX_OK) {
+        return NULL;
+    }
+
     return ngx_sprintf(buf, "%O", r->request_length);
 }
 
@@ -981,23 +1048,41 @@
 
 
 static u_char *
-ngx_http_log_variable(ngx_http_request_t *r, u_char *buf, ngx_http_log_op_t *op)
-{
-    ngx_http_log_loc_conf_t    *lcf;
+ngx_http_log_variable(ngx_http_request_t *r, u_char *buf, u_char *end,
+    ngx_http_log_op_t *op)
+{
+    uintptr_t                   len;
+    ngx_http_log_loc_conf_t     *lcf;
     ngx_http_variable_value_t  *value;
 
     value = ngx_http_get_indexed_variable(r, op->data);
 
     if (value == NULL || value->not_found) {
+        if (ngx_http_log_check_length(r, buf, end, 1) != NGX_OK) {
+            return NULL;
+        }
+
         *buf = '-';
         return buf + 1;
     }
 
     if (value->escape == 0) {
+        if (ngx_http_log_check_length(r, buf, end, value->len) != NGX_OK) {
+            return NULL;
+        }
+
         return ngx_cpymem(buf, value->data, value->len);
 
     } else {
         lcf = ngx_http_get_module_loc_conf(r, ngx_http_log_module);
+        len = ngx_http_log_escape(lcf, NULL, value->data, value->len);
+
+        if (ngx_http_log_check_length(r, buf, end, value->len + len * 3)
+            != NGX_OK)
+        {
+            return NULL;
+        }
+
         return (u_char *) ngx_http_log_escape(lcf, buf, value->data, value->len);
     }
 }
@@ -1091,9 +1176,10 @@
 
 
 static u_char *
-ngx_http_log_json_variable(ngx_http_request_t *r, u_char *buf,
+ngx_http_log_json_variable(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
+    uintptr_t                   len;
     ngx_http_variable_value_t  *value;
 
     value = ngx_http_get_indexed_variable(r, op->data);
@@ -1103,9 +1189,21 @@
     }
 
     if (value->escape == 0) {
+        if (ngx_http_log_check_length(r, buf, end, value->len) != NGX_OK) {
+            return NULL;
+        }
+
         return ngx_cpymem(buf, value->data, value->len);
 
     } else {
+        len = ngx_escape_json(NULL, value->data, value->len);
+
+        if (ngx_http_log_check_length(r, buf, end, value->len + len)
+            != NGX_OK)
+        {
+            return NULL;
+        }
+
         return (u_char *) ngx_escape_json(buf, value->data, value->len);
     }
 }
@@ -1130,7 +1228,7 @@
 
 static u_char *
 ngx_http_log_unescaped_variable(ngx_http_request_t *r, u_char *buf,
-    ngx_http_log_op_t *op)
+    u_char *end, ngx_http_log_op_t *op)
 {
     ngx_http_variable_value_t  *value;
 
@@ -1140,7 +1238,25 @@
         return buf;
     }
 
+    if (ngx_http_log_check_length(r, buf, end, value->len) != NGX_OK) {
+        return NULL;
+    }
+
     return ngx_cpymem(buf, value->data, value->len);
+}
+
+
+static ngx_int_t
+ngx_http_log_check_length(ngx_http_request_t *r, u_char *buf, u_char *end,
+    size_t len)
+{
+    if (end - buf < (ssize_t) len) {
+        ngx_log_error(NGX_LOG_ALERT, r->connection->log, 0,
+                      "no buffer space in log script copy");
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
 }
 
 
--- a/src/stream/ngx_stream_log_module.c
+++ b/src/stream/ngx_stream_log_module.c
@@ -17,7 +17,7 @@
 typedef struct ngx_stream_log_op_s  ngx_stream_log_op_t;
 
 typedef u_char *(*ngx_stream_log_op_run_pt) (ngx_stream_session_t *s,
-    u_char *buf, ngx_stream_log_op_t *op);
+    u_char *buf, u_char *end, ngx_stream_log_op_t *op);
 
 typedef size_t (*ngx_stream_log_op_getlen_pt) (ngx_stream_session_t *s,
     uintptr_t data);
@@ -115,16 +115,19 @@
 static size_t ngx_stream_log_variable_getlen(ngx_stream_session_t *s,
     uintptr_t data);
 static u_char *ngx_stream_log_variable(ngx_stream_session_t *s, u_char *buf,
-    ngx_stream_log_op_t *op);
+    u_char *end, ngx_stream_log_op_t *op);
 static uintptr_t ngx_stream_log_escape(u_char *dst, u_char *src, size_t size);
 static size_t ngx_stream_log_json_variable_getlen(ngx_stream_session_t *s,
     uintptr_t data);
 static u_char *ngx_stream_log_json_variable(ngx_stream_session_t *s,
-    u_char *buf, ngx_stream_log_op_t *op);
+    u_char *buf, u_char *end, ngx_stream_log_op_t *op);
 static size_t ngx_stream_log_unescaped_variable_getlen(ngx_stream_session_t *s,
     uintptr_t data);
 static u_char *ngx_stream_log_unescaped_variable(ngx_stream_session_t *s,
-    u_char *buf, ngx_stream_log_op_t *op);
+    u_char *buf, u_char *end, ngx_stream_log_op_t *op);
+
+static ngx_int_t ngx_stream_log_check_length(ngx_stream_session_t *s,
+    u_char *buf, u_char *end, size_t len);
 
 
 static void *ngx_stream_log_create_main_conf(ngx_conf_t *cf);
@@ -200,7 +203,7 @@
 static ngx_int_t
 ngx_stream_log_handler(ngx_stream_session_t *s)
 {
-    u_char                     *line, *p;
+    u_char                     *line, *p, *end;
     size_t                      len, size;
     ssize_t                     n;
     ngx_str_t                   val;
@@ -257,6 +260,8 @@
             }
         }
 
+        len += NGX_LINEFEED_SIZE;
+
         if (log[l].syslog_peer) {
 
             /* length of syslog's PRI and HEADER message parts */
@@ -267,8 +272,6 @@
             goto alloc_line;
         }
 
-        len += NGX_LINEFEED_SIZE;
-
         buffer = log[l].file ? log[l].file->data : NULL;
 
         if (buffer) {
@@ -284,13 +287,18 @@
             if (len <= (size_t) (buffer->last - buffer->pos)) {
 
                 p = buffer->pos;
+                end = p + len - NGX_LINEFEED_SIZE;
 
                 if (buffer->event && p == buffer->start) {
                     ngx_add_timer(buffer->event, buffer->flush);
                 }
 
-                for (i = 0; i < log[l].format->ops->nelts; i++) {
-                    p = op[i].run(s, p, &op[i]);
+                for (i = 0; i < log[l].format->ops->nelts && p; i++) {
+                    p = op[i].run(s, p, end, &op[i]);
+                }
+
+                if (p == NULL) {
+                    return NGX_ERROR;
                 }
 
                 ngx_linefeed(p);
@@ -313,13 +321,18 @@
         }
 
         p = line;
+        end = p + len - NGX_LINEFEED_SIZE;
 
         if (log[l].syslog_peer) {
             p = ngx_syslog_add_header(log[l].syslog_peer, line);
         }
 
-        for (i = 0; i < log[l].format->ops->nelts; i++) {
-            p = op[i].run(s, p, &op[i]);
+        for (i = 0; i < log[l].format->ops->nelts && p; i++) {
+            p = op[i].run(s, p, end, &op[i]);
+        }
+
+        if (p == NULL) {
+            return NGX_ERROR;
         }
 
         if (log[l].syslog_peer) {
@@ -663,7 +676,7 @@
 
 
 static u_char *
-ngx_stream_log_copy_short(ngx_stream_session_t *s, u_char *buf,
+ngx_stream_log_copy_short(ngx_stream_session_t *s, u_char *buf, u_char *end,
     ngx_stream_log_op_t *op)
 {
     size_t     len;
@@ -671,6 +684,10 @@
 
     len = op->len;
     data = op->data;
+
+    if (ngx_stream_log_check_length(s, buf, end, len) != NGX_OK) {
+        return NULL;
+    }
 
     while (len--) {
         *buf++ = (u_char) (data & 0xff);
@@ -682,9 +699,13 @@
 
 
 static u_char *
-ngx_stream_log_copy_long(ngx_stream_session_t *s, u_char *buf,
+ngx_stream_log_copy_long(ngx_stream_session_t *s, u_char *buf, u_char *end,
     ngx_stream_log_op_t *op)
 {
+    if (ngx_stream_log_check_length(s, buf, end, op->len) != NGX_OK) {
+        return NULL;
+    }
+
     return ngx_cpymem(buf, (u_char *) op->data, op->len);
 }
 
@@ -745,22 +766,39 @@
 
 
 static u_char *
-ngx_stream_log_variable(ngx_stream_session_t *s, u_char *buf,
+ngx_stream_log_variable(ngx_stream_session_t *s, u_char *buf, u_char *end,
     ngx_stream_log_op_t *op)
 {
+    uintptr_t                     len;
     ngx_stream_variable_value_t  *value;
 
     value = ngx_stream_get_indexed_variable(s, op->data);
 
     if (value == NULL || value->not_found) {
+        if (ngx_stream_log_check_length(s, buf, end, 1) != NGX_OK) {
+            return NULL;
+        }
+
         *buf = '-';
         return buf + 1;
     }
 
     if (value->escape == 0) {
+        if (ngx_stream_log_check_length(s, buf, end, value->len) != NGX_OK) {
+            return NULL;
+        }
+
         return ngx_cpymem(buf, value->data, value->len);
 
     } else {
+        len = ngx_stream_log_escape(NULL, value->data, value->len);
+
+        if (ngx_stream_log_check_length(s, buf, end, value->len + len * 3)
+            != NGX_OK)
+        {
+            return NULL;
+        }
+
         return (u_char *) ngx_stream_log_escape(buf, value->data, value->len);
     }
 }
@@ -847,9 +885,10 @@
 
 
 static u_char *
-ngx_stream_log_json_variable(ngx_stream_session_t *s, u_char *buf,
+ngx_stream_log_json_variable(ngx_stream_session_t *s, u_char *buf, u_char *end,
     ngx_stream_log_op_t *op)
 {
+    uintptr_t                     len;
     ngx_stream_variable_value_t  *value;
 
     value = ngx_stream_get_indexed_variable(s, op->data);
@@ -859,9 +898,21 @@
     }
 
     if (value->escape == 0) {
+        if (ngx_stream_log_check_length(s, buf, end, value->len) != NGX_OK) {
+            return NULL;
+        }
+
         return ngx_cpymem(buf, value->data, value->len);
 
     } else {
+        len = ngx_escape_json(NULL, value->data, value->len);
+
+        if (ngx_stream_log_check_length(s, buf, end, value->len + len)
+            != NGX_OK)
+        {
+            return NULL;
+        }
+
         return (u_char *) ngx_escape_json(buf, value->data, value->len);
     }
 }
@@ -887,7 +938,7 @@
 
 static u_char *
 ngx_stream_log_unescaped_variable(ngx_stream_session_t *s, u_char *buf,
-                                  ngx_stream_log_op_t *op)
+    u_char *end, ngx_stream_log_op_t *op)
 {
     ngx_stream_variable_value_t  *value;
 
@@ -897,7 +948,25 @@
         return buf;
     }
 
+    if (ngx_stream_log_check_length(s, buf, end, value->len) != NGX_OK) {
+        return NULL;
+    }
+
     return ngx_cpymem(buf, value->data, value->len);
+}
+
+
+static ngx_int_t
+ngx_stream_log_check_length(ngx_stream_session_t *s, u_char *buf, u_char *end,
+    size_t len)
+{
+    if (end - buf < (ssize_t) len) {
+        ngx_log_error(NGX_LOG_ALERT, s->connection->log, 0,
+                      "no buffer space in log script copy");
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
 }
 
 
